### PR TITLE
Offboarding a Fund Process & Rake Task

### LINKED ISSUE #2634: Figure out how to delete an instance
Thanks for creating an issue! Please fill out this form so we can be sure to have all the information we need, and to minimize back and forth.

* **What are we trying to do?**

There's a high chance that we'll be dropping several clients before the end of the calendar year, and I don't want sensitive data to hang around for any longer than it needs to. I would like to write a playbook for dropping a fund from a software perspective, so that when someone asks we'll be ready.

I think this will probably happen around September at the earliest so having something by then is good.

* **What feature or behavior is this required for?**

fund management and administration

* **How could we solve this issue? (Not knowing is okay!)**

this should cover the 'delete rows associated with a fund, including relations and audit trails' parts. Probably not sessions. Probably should be run very very very manually and have some kind of safety switch.

* **Anything else?**

@lomky can you handle this one?

### PR DESCRIPTION
**Pull request by lomky**
I rule and have completed some work on Case Manager that's ready for review!
Adds first draft of Offboarding a Fund

This pull request makes the following changes:
* Teaches the Fund object how to delete its related Patient and Administrative data
* Adds a rake task that invokes those functions & deletes the PaperTrails for a fund
* Adds the document

It relates to the following issue #s: 
* Fixes #2634 

For reviewer:
* Adjust the title to explain what it does for the notification email to the listserv.
* Tag this PR:
  * `feature` if it contains a feature, fix, or similar. This is anything that contains a user-facing fix in some way, such as frontend changes, alterations to backend behavior, or bug fixes.
  * `dependencies` if it contains library upgrades or similar. This is anything that upgrades any dependency, such as a Gemfile update or npm package upgrade.
* If it contains neither, no need to tag this PR.

### CODE REVIEW

File: docs/administering/OFFBOARDING_A_FUND.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+#### Remove the DNS in Namecheap
+  -  Open up the Namecheap control panel and head to Advanced DNS.
+  -  Remove the CNAME Record for the host 
+#### Remove the subdomain in heroku
+  -  Open up the pipeline in Heroku and go to the production instance, then open up Settings, then go to Domains.
+  -  Click the edit icon for the fund's domain and select Remove Domain
+  - Remove the domain's certs, if not done automatically
+#### Remove defunct manual database backups in heroku
+  -  Open up the pipeline in Heroku and go to the production instance, then open up Resources, then click on Heroku Postgres, then open Durability
+  - Remove any Manual Backups & Data Exports
```
Comment (lomky) [inline docs/administering/OFFBOARDING_A_FUND.md:150]: We have a few of these hanging out, all 6-9 months old. I think we have to have this step to ensure we won't retain copies. Are there any caveats we want to add for technical needs?

File: docs/administering/OFFBOARDING_A_FUND.md [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+### Remaining Fund Information
+
+#### Sentry
+Sentry errors for the fund, containing no PII, may be retained for up to 90 days.
+
+#### Database backups
+Automatic database backups are retained for four days.
+In the event that we need to rollback the database we will repeat the removal of all fund information.
```
Comment (lomky) [inline docs/administering/OFFBOARDING_A_FUND.md:159]: Sound about right?

File: docs/administering/OFFBOARDING_A_FUND.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+  starting_counts[fund.name] = fund_counts
+end
+fund_counts = {}
+models_fund_agnostic.each do |model|
+  fund_counts[model.to_s] = model.all.size
+end
+starting_counts['fund_agnostic'] = fund_counts
+
+#save starting counts to file
+$stdout = File.new('initail_state.out', 'w')
```
Comment (colinxfleming) [inline docs/administering/OFFBOARDING_A_FUND.md:50]: spelling
```suggestion
$stdout = File.new('initial_state.out', 'w')
```

File: docs/administering/OFFBOARDING_A_FUND.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+exit
+```
+
+diff those files!
+
+  - The other fund objects should not have changed significantly
+  - There should be a jump in PaperTrails corresponding to the total object count we've deleted
+
+`diff -y initial_count.out post_deletion.out`
```
Comment (colinxfleming) [inline docs/administering/OFFBOARDING_A_FUND.md:110]: I think these commands aren't plopping the right stuff in the files, though maybe I have something configured weird?

File: docs/administering/OFFBOARDING_A_FUND.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+Remove the PaperTrails associated with the fund, and finally the Fund row itself.
+
+run `rails c` again
+
+```ruby
+# delete the PaperTrails
+#  🚨 ⚠️ 🚨 ⚠️ 🚨 This is the point of no return.
+# Once these are deleted, we cannot restore the funds information from this database
+defunct_fund_versions = PaperTrail::Version.where_object(fund_id: id)
```
Comment (colinxfleming) [inline docs/administering/OFFBOARDING_A_FUND.md:122]: need to redeclare fund_id here if we're going to pop in and out of the console

File: docs/administering/OFFBOARDING_A_FUND.md (hunk 1 above)
Comment (colinxfleming) [reply docs/administering/OFFBOARDING_A_FUND.md:152]: nah I think this is a good call, though it does make me think 'maybe start this process by taking a db snapshot' is a good idea?

**Review by colinxfleming — COMMENTED** [review]
approach seems sound to me! couple rough edges to send off in this pass around the QA, but the premise is sound and I think the destroys are all there

File: docs/administering/OFFBOARDING_A_FUND.md (hunk 4 above)
Comment (lomky) [reply docs/administering/OFFBOARDING_A_FUND.md:110]: what are you getting? I see a structure like:

```ruby
{
  "fund a" => {
              "Notes" => 400
              "Patient" => 500
               ...
              }
   "fund b" => { 
             "Notes" => 20
              "Patient" => 40
               ...
               }
     "fund_agnostic" => {
          "Funds" => 4
          "Sessions" => 500
          "PaperTrail" => 40904
}
```

File: docs/administering/OFFBOARDING_A_FUND.md (hunk 5 above)
Comment (lomky) [reply docs/administering/OFFBOARDING_A_FUND.md:122]:
```suggestion
defunct_fund_versions = PaperTrail::Version.where_object(fund_id: defunct_fund_id)
```

File: docs/administering/OFFBOARDING_A_FUND.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+## Main steps
+
+At a high level, we are:
+
+  - Removing Fund-specific rows from our production database
+  - Removing the fund configuration on our infrastructure
+  - Removing the fund from our contact systems
+
+### Removing Production Data for the Fund
+
```
Comment (lomky) [inline docs/administering/OFFBOARDING_A_FUND.md:16]:
```suggestion

Start by taking a snapshot of the database in heroku:
  - Open up the pipeline in Heroku and go to the production instance, then open up Resources, then click on Heroku Postgres, then open Durability
  - Create a manual snapshot
  
```

File: docs/administering/OFFBOARDING_A_FUND.md [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+Start by taking a snapshot of the database in heroku:
+  - Open up the pipeline in Heroku and go to the production instance, then open up Resources, then click on Heroku Postgres, then open Durability
+  - Create a manual snapshot
+  
+I suggest doing this via the Heroku CLI, as I have found the GUI console can close unexpectedly. Heroku run instances are ephemeral and will not retain a saved file from one run to the next.
+
+run `bash`. You should stay in this instance throughout or the reference files won't be retained.
+
+Once connected, run `rails c`
```
Comment (lomky) [inline docs/administering/OFFBOARDING_A_FUND.md:25]: TODO: Convert to a rake script once we're happy with this

File: docs/administering/OFFBOARDING_A_FUND.md [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+Sentry errors for the fund, containing no PII, may be retained for up to 90 days.
+
+#### Database backups
+Automatic database backups are retained for four days.
+
+In the event that we need to rollback the database we will repeat the removal of all fund information.
+
+### Removing the fund from our contact systems
+
+  - Remove the fund from our Fund Information & Point of Contact document
```
Comment (colinxfleming) [inline docs/administering/OFFBOARDING_A_FUND.md:161]: I think we can leave em on this and just mark the row defunct -- I can think of good reasons that we'd want to continue to be in touch that would outweigh risk imo

**Review by colinxfleming — COMMENTED** [review]
I dry ran this and it works great! I think ready for conversion to the rake task. nicely done

File: docs/administering/OFFBOARDING_A_FUND.md [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+
+Copy any exerpts from the output files you want for later reference and `exit` the bash
+
+### Removing the Subdomain Configuration
+
+#### Remove Google OAuth
+  - Go to our [GCP project](https://console.cloud.google.com/apis/credentials?project=daria-services-multitenant)
+  - Open up DARIA Services Production from the list of Client IDs to edit
+  - Remove the Authorized redirect URI for the fund's subdomain
+#### Remove the DNS in Namecheap
```
Comment (xmunoz) [inline docs/administering/OFFBOARDING_A_FUND.md:138]:
```suggestion
#### Remove the DNS in Cloudflare
```

File: docs/administering/OFFBOARDING_A_FUND.md [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+  - Open up the Namecheap control panel and head to Advanced DNS.
```
Comment (xmunoz) [inline docs/administering/OFFBOARDING_A_FUND.md:139]:
```suggestion
  - Open up the Cloudflare control panel and head to [DNS](https://dash.cloudflare.com/04a3c6a398793e6f59a1dbcadcd9ddc9/dariaservices.com/dns).
```

**Review by xmunoz — APPROVED** [review]
Add the suggested updates to switch out Namecheap with Cloudflare.

**Review by xmunoz — APPROVED** [review]
LGTM

**Review by colinxfleming — APPROVED** [review]
thank you @lomky !

**Pull request merged**